### ISSUE TITLE
[Feature Request] Make sharing more automatic with sharing profiles

### ISSUE THREAD

**Issue by To-om** [TheHive-Project/TheHive#2121]
### Request Type

Feature Request

### Feature Description

Make sharing easier for users by defining sharing profiles between organisations. Administrators can define rules that share automatically cases, tasks and observables to other organisations. 

#### Organisation links
Links between two organisations contain the property `linkType` (String).
```
PUT /api/v0/organisation/<id1>/link/<id2>
 - linkType: Option[String] (default=default) // link type between org_id1 -> org_id2
 - otherLinkType: Option[String] (default=default) // link type between org_id2 -> org_id1
PUT /api/v0/organisation/<id>/links
 - organisations: Seq[String]
 - linkType: Option[String] (default=default) // link type between org_id -> other organisations
 - otherLinkType: Option[String] (default=default) // link type between other organisations -> org_id
```

The type of link contain the name of the `SharingProfile` which is defined by:
```
 - name: String
 - description: String
 - autoShare: Boolean // share automatically new case ?
 - editable: Boolean // if editable the behaviour can be overridden by `SharingParameter`
 - permissionProfile: String
 - taskRule: String // rule applied to automatic share
 - observableRule: String
```

The list of available `SharingProfile`s can be obtain with:
```
GET /api/v1/sharingProfile
```

#### Sharing rules
Sharing rules defines which tasks/observables are automatically shared to other organisations which has already access to the current case.
 - `manual` (default): nothing is automatically shared.
 - `existingOnly`: share all elements when applied. New element won't be shared automatically.
 - `upcomingOnly`: only new elements are shared.
 - `all`: all elements are shares, including new ones.

#### Organisation property
An organisation contains the default sharing rules applied to case created in or manually shared to this organisation.
```
POST /api/v1/organisation
 - name: String
 - description: String
 - taskRule: Option[String] (default=manual)
 - observableRule: Option[String] (default=manual)
PATCH /api/v1/organisation/<id>
 - name: String
 - description: String
 - taskRule: Option[String] (default=manual)
 - observableRule: Option[String] (default=manual)
```

`taskRule` and `organisationRule` has been added in organisation outputs (both in APIv0 and APIv1).

**[BREAKING CHANGE]**
The links in organisation outputs in APIv1 (APIv0 is unchanged) include organisation link types. The property `links` (was `Seq[String]`) becomes `Seq[OrganisationLink]`:
```
link: [
  {
    toOrganisation: String
    linkType: String
    otherLinkType: String
  }
]
```

#### Create new case
A new case is automatically share according to organisation links (`autoShare`=true). If permitted (`editable`=true), sharing profiles can be overridden and extended with the parameter `sharingParameter`.

If not provided, the sharing rules (`taskRule` and `observableRule`) used is those which are in user organisation. 
```
POST /api/v0/case
 - _: Case
 - tasks: Seq[Task]
 - caseTemplate: Option[String]
 - sharingParameters: Seq[InputShare]
 - taskRule: Option[String]
 - observableRule: Option[String]
```

#### Case property
The sharing rules (`taskRule` and `observableRule`) indicates what is automatically shared to viewing organisations. They are stored in the case share (related to one organisation).
They can be updated with a PATCH on the case.
```
PATCH /api/v0/case/<id>
 - taskRule: String
 - observableRule: String
PATCH /api/v1/case/<id>
 - taskRule: String
 - observableRule: String
```

The sharing rule for organisations other that the current user's one is not possible.

#### Create case from alert
A case, created from an alert, is automatically shared according to organisation links. If permitted (`editable`=true), sharing profiles can be overridden and extended with the parameter `sharingParameter`.

The tasks and the observables are shared according to sharing rules configured on the user organisation. These rules can be overriden by `taskRule` and `observableRule`.
```
POST /api/v0/alert/<id>/createCase
 - caseTemplate: Option[String]
 - sharingParameters: Seq[InputShare]
 - taskRule: Option[String]
 - observableRule: Option[String]
POST /api/v1/alert/<id>/createCase
 - caseTemplate: Option[String]
 - sharingParameters: Seq[InputShare]
 - taskRule: Option[String]
 - observableRule: Option[String]
```

#### Share a case
**[BREAKING CHANGE]**
When a case is shared, the sharing rules used for tasks and observables are those which come from the current share (linked to user organisation). They can't be overridden. The rules in the parameter `shares` are used for the created share.
Sharing a case shares automatically the tasks and observables according to rules in current share.  
```
POST /api/v0/case/<id>/shares
 - shares: Seq[InputShare]
POST /api/v1/case/<id>/shares
 - shares: Seq[InputShare]
```

An `InputShare` is defined as follow:
```
  - organisation: String
  - share: Option[Boolean] (default=true)
  - profile: Option[String] (default=Analyst)
  - taskRule: Option[String] (default=manual)
  - observableRule: Option[String] (default=manual)
```

**Comment by vdebergue** [comment]
For the "Sharing rules" it might be good and really powerful to have rules based on `pap` or `tlp`. 
For instance:  "share only observables (existing and new) with `pap < Amber`"

**Comment by To-om** [comment]
I agree.  Sharing rules will be extended, maybe not for this release.

**Comment by vdebergue** [comment]
So maybe use a json object for the rules instead of a string so that the api will not break in the future ?
